diff --git a/libavcodec/ccaption_dec.c b/libavcodec/ccaption_dec.c
index 6f4d4fca41..75407d0614 100644
--- a/libavcodec/ccaption_dec.c
+++ b/libavcodec/ccaption_dec.c
@@ -868,8 +868,10 @@ static int decode(AVCodecContext *avctx, AVSubtitle *sub,
     for (i = 0; i < len; i += 3) {
         uint8_t hi, cc_type = bptr[i] & 1;
 
-        if (ctx->data_field < 0)
-            ctx->data_field = cc_type;
+        av_log(avctx, AV_LOG_INFO, "CC608 RAW type=%d data=%02x %02x\n", cc_type, bptr[i + 1], bptr[i + 2]);
+
+        //if (ctx->data_field < 0)
+        //    ctx->data_field = cc_type;
 
         if (validate_cc_data_pair(bptr + i, &hi))
             continue;
