diff --git a/demux/demux.c b/demux/demux.c
index d1accba789..0d0d894748 100644
--- a/demux/demux.c
+++ b/demux/demux.c
@@ -49,6 +49,14 @@
 #include "stheader.h"
 #include "cue.h"
 
+#ifdef __ANDROID__
+#include <android/log.h>
+#define CCEMIT(...) __android_log_print(ANDROID_LOG_ERROR, "CC608Log", __VA_ARGS__)
+#else
+#define CCEMIT(...) do {} while (0)
+#endif
+
+
 // Demuxer list
 extern const demuxer_desc_t demuxer_desc_mpv;
 extern const demuxer_desc_t demuxer_desc_edl;
@@ -63,6 +71,9 @@ extern const demuxer_desc_t demuxer_desc_directory;
 extern const demuxer_desc_t demuxer_desc_disc;
 extern const demuxer_desc_t demuxer_desc_rar;
 extern const demuxer_desc_t demuxer_desc_libarchive;
+#if HAVE_SUBRANDR
+extern const demuxer_desc_t demuxer_desc_sbr;
+#endif
 extern const demuxer_desc_t demuxer_desc_null;
 extern const demuxer_desc_t demuxer_desc_timeline;
 
@@ -76,6 +87,9 @@ static const demuxer_desc_t *const demuxer_list[] = {
     &demuxer_desc_matroska,
 #if HAVE_LIBARCHIVE
     &demuxer_desc_libarchive,
+#endif
+#if HAVE_SUBRANDR
+    &demuxer_desc_sbr,
 #endif
     &demuxer_desc_lavf,
     &demuxer_desc_mf,
@@ -1207,7 +1221,7 @@ void demux_stop_thread(struct demuxer *demuxer)
         mp_cond_signal(&in->wakeup);
         mp_mutex_unlock(&in->lock);
         mp_thread_join(in->thread);
-        in->threading = false;
+        in->threading = false; 
         in->thread_terminate = false;
     }
 }
@@ -1242,16 +1256,21 @@ const char *stream_type_name(enum stream_type type)
     default:            return "unknown";
     }
 }
-
+ 
 static struct sh_stream *demuxer_get_cc_track_locked(struct sh_stream *stream)
 {
+      CCEMIT("QQQ CCDBG demuxer_get_cc_track_locked: stream=%p type=%d codec=%s ds->cc=%p", stream, stream->type, stream->codec && stream->codec->codec ? stream->codec->codec : "null",   stream->ds->cc);
+           
     struct sh_stream *sh = stream->ds->cc;
 
     if (!sh) {
         sh = demux_alloc_sh_stream(STREAM_SUB);
         if (!sh)
             return NULL;
+        CCEMIT("QQQ CCDBG created CC track: sh=%p type=%d codec=%s default=%d",  sh, sh->type, sh->codec && sh->codec->codec ? sh->codec->codec : "null",  sh->default_track);
+       
         sh->codec->codec = "eia_608";
+        //sh->codec->codec = "ccaption";
         sh->default_track = true;
         sh->hls_bitrate = stream->hls_bitrate;
         sh->program_id = stream->program_id;
@@ -1265,11 +1284,18 @@ static struct sh_stream *demuxer_get_cc_track_locked(struct sh_stream *stream)
 
 void demuxer_feed_caption(struct sh_stream *stream, demux_packet_t *dp)
 {
+    CCEMIT( "QQQ CCDBG demuxer_feed_caption: src_stream=%p src_type=%d src_codec=%s dp=%p pts=%f dts=%f len=%d",  stream, stream->type,  stream->codec && stream->codec->codec ? stream->codec->codec : "null",   dp, dp ? dp->pts : -9999.0, dp ? dp->dts : -9999.0, dp ? dp->len : -1);
+       
     struct demux_internal *in = stream->ds->in;
 
+    CCEMIT( "QQQ CCDBG demuxer_feed_caption1");
+
     mp_mutex_lock(&in->lock);
     struct sh_stream *sh = demuxer_get_cc_track_locked(stream);
-    if (!sh) {
+
+    CCEMIT("QQQ CCDBG demuxer_feed_caption2: cc_track=%p", sh);
+    
+    if (!sh) { 
         mp_mutex_unlock(&in->lock);
         talloc_free(dp);
         return;
@@ -1278,6 +1304,7 @@ void demuxer_feed_caption(struct sh_stream *stream, demux_packet_t *dp)
     dp->keyframe = true;
     dp->pts = MP_ADD_PTS(dp->pts, -in->ts_offset);
     dp->dts = MP_ADD_PTS(dp->dts, -in->ts_offset);
+    CCEMIT( "QQQ CCDBG demuxer_feed_caption adjusted: pts=%f keyframe=%d",   dp->pts, dp->keyframe);    
     dp->stream = sh->index;
     add_packet_locked(sh, dp);
     mp_mutex_unlock(&in->lock);
diff --git a/filters/f_decoder_wrapper.c b/filters/f_decoder_wrapper.c
index 5e2e8d74c0..f429359ffc 100644
--- a/filters/f_decoder_wrapper.c
+++ b/filters/f_decoder_wrapper.c
@@ -52,6 +52,14 @@
 #include "f_demux_in.h"
 #include "filter_internal.h"
 
+#ifdef __ANDROID__
+#include <android/log.h>
+#define CCEMIT(...) __android_log_print(ANDROID_LOG_ERROR, "CC608Log", __VA_ARGS__)
+#else
+#define CCEMIT(...) do {} while (0)
+#endif
+
+
 struct dec_queue_opts {
     bool use_queue;
     int64_t max_bytes;
@@ -455,8 +463,8 @@ static bool reinit_decoder(struct priv *p)
 
         p->decoder = driver->create(p->decf, p->codec, sel->decoder);
         if (p->decoder) {
-            p->codec->decoder = talloc_strdup(p, sel->decoder);
-            p->codec->decoder_desc = talloc_strdup(p, sel->desc && sel->desc[0] ? sel->desc : NULL);
+            p->codec->decoder = talloc_strdup(p->codec, sel->decoder);
+            p->codec->decoder_desc = talloc_strdup(p->codec, sel->desc && sel->desc[0] ? sel->desc : NULL);
             MP_VERBOSE(p, "Selected decoder: %s", sel->decoder);
             if (p->codec->decoder_desc)
                 MP_VERBOSE(p, " - %s", p->codec->decoder_desc);
@@ -723,12 +731,17 @@ static bool process_decoded_frame(struct priv *p, struct mp_frame *frame)
 
         crazy_video_pts_stuff(p, mpi);
 
-        struct demux_packet *ccpkt = new_demux_packet_from_buf(p->public.f->packet_pool,
-                                                               mpi->a53_cc);
+        //CCEMIT("QQQ CCDBG  process_decoded_frame: pts=%f dts=%f a53_cc=%p",  mpi->pts, mpi->dts, mpi->a53_cc);
+
+        struct demux_packet *ccpkt = new_demux_packet_from_buf(p->public.f->packet_pool, mpi->a53_cc);
+ 
+        //CCEMIT("QQQ CCDBG new_demux_packet_from_buf: ccpkt=%p", ccpkt);
+
         if (ccpkt) {
             av_buffer_unref(&mpi->a53_cc);
             ccpkt->pts = mpi->pts;
             ccpkt->dts = mpi->dts;
+            CCEMIT("QQQ CCDBG feeding caption: pts=%f dts=%f len=%d\n",  ccpkt->pts, ccpkt->dts, ccpkt->len);
             demuxer_feed_caption(p->header, ccpkt);
         }
 
diff --git a/sub/dec_sub.c b/sub/dec_sub.c
index 118221a658..b955ebebcb 100644
--- a/sub/dec_sub.c
+++ b/sub/dec_sub.c
@@ -34,11 +34,25 @@
 #include "misc/dispatch.h"
 #include "osdep/threads.h"
 
+#ifdef __ANDROID__
+#include <android/log.h>
+#define CCEMIT(...) __android_log_print(ANDROID_LOG_ERROR, "CC608Log", __VA_ARGS__)
+#else
+#define CCEMIT(...) do {} while (0)
+#endif
+
+
 extern const struct sd_functions sd_ass;
 extern const struct sd_functions sd_lavc;
+#if HAVE_SUBRANDR
+extern const struct sd_functions sd_sbr;
+#endif
 
 static const struct sd_functions *const sd_list[] = {
     &sd_lavc,
+#if HAVE_SUBRANDR
+    &sd_sbr,
+#endif
     &sd_ass,
     NULL
 };
@@ -69,6 +83,7 @@ struct dec_sub {
 
     struct mp_codec_params *codec;
     double start, end;
+    char *lang;
 
     double last_vo_pts;
     struct sd *sd;
@@ -135,7 +150,7 @@ static void destroy_cached_pkts(struct dec_sub *sub)
     }
     sub->cached_pkt_pos = 0;
     sub->num_cached_pkts = 0;
-}
+} 
 
 void sub_destroy(struct dec_sub *sub)
 {
@@ -153,6 +168,8 @@ void sub_destroy(struct dec_sub *sub)
 
 static struct sd *init_decoder(struct dec_sub *sub)
 {
+    CCEMIT("QQQ CCDBG init_decoder(sub): codec=%s",  sub->codec && sub->codec->codec ? sub->codec->codec : "null");
+       
     for (int n = 0; sd_list[n]; n++) {
         const struct sd_functions *driver = sd_list[n];
         struct sd *sd = talloc(NULL, struct sd);
@@ -168,17 +185,21 @@ static struct sd *init_decoder(struct dec_sub *sub)
             .preload_ok = true,
         };
 
-        if (sd->driver->init(sd) >= 0)
+      CCEMIT("QQQ CCDBG trying subtitle driver=%s codec=%s",  driver->name, sub->codec && sub->codec->codec ? sub->codec->codec : "null");
+       
+
+        if (sd->driver->init(sd) >= 0) {
+          CCEMIT( "QQQ CCDBG subtitle driver selected=%s", driver->name); 
             return sd;
+        }
 
         talloc_free(sd);
     }
 
-    MP_ERR(sub, "Could not find subtitle decoder for format '%s'.\n",
-           sub->codec->codec);
+    CCEMIT( "Could not find subtitle decoder for format '%s'.",  sub->codec->codec);
     return NULL;
 }
-
+ 
 // Thread-safety of the returned object: all functions are thread-safe,
 // except sub_get_bitmaps() and sub_get_text(). Decoder backends (sd_*)
 // do not need to acquire locks.
@@ -187,6 +208,15 @@ static struct sd *init_decoder(struct dec_sub *sub)
 struct dec_sub *sub_create(struct mpv_global *global, struct track *track,
                            struct attachment_list *attachments, int order)
 {
+    CCEMIT( "QQQ CCDBG sub_create: track=%p type=%d user_tid=%d demuxer_id=%d ff_index=%d codec=%s lang=%s\n",
+           track, track ? track->type : -1,
+           track ? track->user_tid : -1,
+           track ? track->demuxer_id : -1,
+           track ? track->ff_index : -1,
+           track && track->stream && track->stream->codec && track->stream->codec->codec
+                ? track->stream->codec->codec : "null",
+           track && track->lang ? track->lang : "null");
+       
     mp_assert(track->stream && track->stream->type == STREAM_SUB);
 
     struct dec_sub *sub = talloc(NULL, struct dec_sub);
@@ -198,6 +228,7 @@ struct dec_sub *sub_create(struct mpv_global *global, struct track *track,
         .shared_opts_cache = m_config_cache_alloc(sub, global, &mp_subtitle_shared_sub_opts),
         .sh = track->stream,
         .codec = track->stream->codec,
+        .lang = track->lang,
         .attachments = talloc_steal(sub, attachments),
         .play_dir = 1,
         .order = order,
@@ -210,7 +241,13 @@ struct dec_sub *sub_create(struct mpv_global *global, struct track *track,
     sub->shared_opts = sub->shared_opts_cache->opts;
     mp_mutex_init(&sub->lock);
 
+    CCEMIT("QQQ CCDBG sub_create before init_decoder: codec=%s\n",
+       sub->codec && sub->codec->codec ? sub->codec->codec : "null");
+       
     sub->sd = init_decoder(sub);
+    
+    CCEMIT( "QQQ CCDBG sub_create after init_decoder: sd=%p\n", sub->sd);    
+    
     if (sub->sd) {
         update_subtitle_speed(sub);
         return sub;
diff --git a/sub/sd_lavc.c b/sub/sd_lavc.c
index 5cbc6d0a21..be2688fbc3 100644
--- a/sub/sd_lavc.c
+++ b/sub/sd_lavc.c
@@ -37,6 +37,14 @@
 
 #define MAX_QUEUE 4
 
+#ifdef __ANDROID__
+#include <android/log.h>
+#define CCEMIT(...) __android_log_print(ANDROID_LOG_ERROR, "CC608Log", __VA_ARGS__)
+#else
+#define CCEMIT(...) do {} while (0)
+#endif
+
+
 struct sub {
     bool valid;
     AVSubtitle avsub;
@@ -75,9 +83,20 @@ struct sd_lavc_priv {
 
 static int init(struct sd *sd)
 {
-    enum AVCodecID cid = mp_codec_to_av_codec_id(sd->codec->codec);
+    const char *codec_name = sd->codec->codec;
+    if (codec_name && strcmp(codec_name, "eia_608") == 0)
+        codec_name = "ccaption";
+
+    CCEMIT("QQQ MAP codec_name=%s orig=%s\n",
+           codec_name ? codec_name : "null",
+           sd->codec && sd->codec->codec ? sd->codec->codec : "null");
+
+    enum AVCodecID cid = mp_codec_to_av_codec_id(codec_name);
+
+    CCEMIT("QQQ MAP cid=%d\n", cid);
+    
+    enum AVCodecID cid = mp_codec_to_av_codec_id(codec_name);
 
-    // Supported codecs must be known to decode to paletted bitmaps
     switch (cid) {
     case AV_CODEC_ID_DVB_SUBTITLE:
     case AV_CODEC_ID_DVB_TELETEXT:
@@ -85,6 +104,7 @@ static int init(struct sd *sd)
     case AV_CODEC_ID_XSUB:
     case AV_CODEC_ID_DVD_SUBTITLE:
     case AV_CODEC_ID_ARIB_CAPTION:
+    case AV_CODEC_ID_EIA_608:
         break;
     default:
         return -1;
@@ -121,7 +141,7 @@ static int init(struct sd *sd)
     }
     }
 
-    MP_VERBOSE(sd, "Using subtitle decoder %s\n", sub_codec->name);
+    MP_VERBOSE(sd, "Using subtitle decoder %s\n", sub_codec->name); 
 
     priv->avpkt = av_packet_alloc();
     priv->codec = sd->codec;
@@ -129,10 +149,16 @@ static int init(struct sd *sd)
         goto error;
     if (mp_set_avctx_codec_headers(ctx, sd->codec) < 0)
         goto error;
-    priv->pkt_timebase = mp_get_codec_timebase(sd->codec);
+    priv->pkt_timebase = mp_get_codec_timebase(sd->codec); 
     ctx->pkt_timebase = priv->pkt_timebase;
+    CCEMIT("QQQ CCDBG sd_lavc about to open subtitle decoder: codec=%s id=%d", sub_codec ? sub_codec->name : "null",
+       sub_codec ? sub_codec->id : -1);
+       
     if (avcodec_open2(ctx, sub_codec, NULL) < 0)
         goto error;
+      
+    CCEMIT("QQQ CCDBG sd_lavc opened subtitle decoder: codec=%s\n",    sub_codec ? sub_codec->name : "null");
+       
     priv->avctx = ctx;
     sd->priv = priv;
     priv->displayed_id = -1;
