diff --git a/filters/f_decoder_wrapper.c b/filters/f_decoder_wrapper.c
index 5e2e8d74c0..cf3eb2a070 100644
--- a/filters/f_decoder_wrapper.c
+++ b/filters/f_decoder_wrapper.c
@@ -52,6 +52,13 @@
 #include "f_demux_in.h"
 #include "filter_internal.h"
 
+#ifdef __ANDROID__
+#include <android/log.h>
+#define CCEMIT(...) __android_log_print(ANDROID_LOG_ERROR, "CC608Log", __VA_ARGS__)
+#else
+#define CCEMIT(...) do {} while (0)
+#endif
+
 struct dec_queue_opts {
     bool use_queue;
     int64_t max_bytes;
@@ -713,7 +720,7 @@ static bool process_decoded_frame(struct priv *p, struct mp_frame *frame)
         // if we were just draining current segment, don't propagate EOF
         if (p->new_segment)
             mp_frame_unref(frame);
-        return true;
+        return true; 
     }
 
     bool segment_ended = false;
@@ -723,12 +730,17 @@ static bool process_decoded_frame(struct priv *p, struct mp_frame *frame)
 
         crazy_video_pts_stuff(p, mpi);
 
-        struct demux_packet *ccpkt = new_demux_packet_from_buf(p->public.f->packet_pool,
-                                                               mpi->a53_cc);
+        CCEMIT("QQQ CCDBG  process_decoded_frame: pts=%f dts=%f a53_cc=%p",  mpi->pts, mpi->dts, mpi->a53_cc);
+
+        struct demux_packet *ccpkt = new_demux_packet_from_buf(p->public.f->packet_pool, mpi->a53_cc);
+ 
+        CCEMIT("QQQ CCDBG new_demux_packet_from_buf: ccpkt=%p", ccpkt);
+
         if (ccpkt) {
             av_buffer_unref(&mpi->a53_cc);
             ccpkt->pts = mpi->pts;
             ccpkt->dts = mpi->dts;
+            CCEMIT("QQQ CCDBG feeding caption: pts=%f dts=%f len=%d\n",  ccpkt->pts, ccpkt->dts, ccpkt->len);            
             demuxer_feed_caption(p->header, ccpkt);
         }
 
